### DISCUSSION TITLE
[MAC] Copying image from clipboard

### DISCUSSION THREAD

**Discussion by guanzhangrtk** [FreeRDP/FreeRDP#7111] [Q&A] [answered]
Hi, I'm trying to copy an image from clipboard but it doesn't seem to be working. Running FreeRDP 2.3.2 on macOS Big Sur 11.3.1 (installed via homebrew).

On the remote Windows 10 host I would copy an image to clipboard (eg. via snip & sketch tool) and when I try to paste remotely (Windows 10) or locally (macOS) they both didn't work. If I start `xfreerdp` with `-clipboard` option then pasting remotely works.

According to #6699 "file" copy to/from clipboard is supposed to be working since 2.3.0 and I'm running 2.3.2 so I expect it to work, unless I'm doing something wrong or this is a limitation of XQuartz 2.8.1. Appreciate any help in advance!

**Comment by akallabeth** [comment, marked as answer]
`xfreerdp` and mac os do not mix well (the X11 server on mac is lacking)

**Reply by guanzhangrtk** [reply]
> `xfreerdp` and mac os do not mix well (the X11 server on mac is lacking)

I see, that's unfortunate. I've just switched from Microsoft's Remote Desktop to xfreerdp due to some issues, and it's working better now. However there are some minor features missing. Do you think this is something the XQuartz devs could possibly add or probably not? Thanks for the response!

**Reply by kimcoldwater** [reply]
Hello.
It's been a while, but I understand that this file-clipboard has been implemented yet.
Has it already been implemented or will it be implemented?

**Reply by mfleisz** [reply]
Sorry - support for file clipboard has not been added to the Mac client yet.

**Reply by hablutzel1** [reply]
Royal TSX, which uses FreeRDP, does have support for file clipboard in macOS. Have they implemented something of their own on top of FreeRDP?

**Reply by robster7674** [reply]
@hablutzel1 no, Royal TSX also doesn't work for images from clipboard. I just reported this Sonnet analysis to them so perhaps they can fix it in their FreeRDP based plugin:

The log confirms the exact failure point:
The message No synthesizer for format image/bmp [0x0000c003] --> CF_UNICODETEXT is FreeRDP's own winpr clipboard subsystem logging that it received image data from xrdp and had nowhere to put it. xrdp is blameless.

The source code confirms why:
mac_cliprdr_server_format_data_response in Clipboard.m has exactly one if block, and it only handles CF_TEXT / CF_OEMTEXT / CF_UNICODETEXT. There is no image branch. Not stubbed out, not disabled — simply never written.

Bonus find:
there are also two == vs = typos in mac_cliprdr_server_format_list that would cause silent text clipboard failures if a server doesn't advertise CF_UNICODETEXT specifically (uses CF_TEXT or CF_OEMTEXT only). Worth including in the same report since it's the same file and same author era.

Hope this helps.
